### PR TITLE
#3432: Raise syntax error on invalid shorthand assignment

### PR DESCRIPTION
**Pull request by 0xe**
**Description:**

Handles invalid shorthand property initializers like `({ a = 1});`
https://github.com/tc39/test262/blob/main/test/language/expressions/object/cover-initialized-name.js

```
error: assignment property is invalid syntax
 --> <anon>:3:4
  |
3 | ({ a = 1 });
  |    ^
```

**BREAKING CHANGE:**

**Related issue (if exists):**
Closes https://github.com/swc-project/swc/issues/3432

### CODE REVIEW

File: crates/swc_ecma_parser/src/parser/object.rs [hunk 1]
```diff
@@ -248,6 +248,7 @@ impl<I: Tokens> ParseObject<Box<Expr>> for Parser<I> {
 
             if eat!(self, '=') {
                 let value = self.include_in_expr(true).parse_assignment_expr()?;
+
```
Comment (0xe) [inline crates/swc_ecma_parser/src/parser/object.rs:251]: will remove.

File: crates/swc_ecma_parser/src/parser/expr.rs [hunk 2]
```diff
@@ -785,6 +785,23 @@ impl<'a, I: Tokens> Parser<I> {
                 }
             }
             return Ok(Box::new(Expr::Arrow(arrow_expr)));
+        } else {
+            // If there's no arrow function, we have to check there's no
+            // AssignProp in lhs to check against assignment in object literals
+            // like (a, {b = 1});
+            for expr_or_spread in paren_items.iter().cloned() {
```
Comment (kdy1) [inline crates/swc_ecma_parser/src/parser/expr.rs:792]: `cloned()` is bad for performance, as it clones the whole content.

File: crates/swc_ecma_parser/src/parser/expr.rs [hunk 3]
```diff
@@ -785,6 +785,23 @@ impl<'a, I: Tokens> Parser<I> {
...
+                if let PatOrExprOrSpread::ExprOrSpread(e) = expr_or_spread {
+                    if let Expr::Object(o) = *e.expr {
+                        for p in o.props.into_iter() {
```
Comment (kdy1) [inline crates/swc_ecma_parser/src/parser/expr.rs:795]: I think `iter()` is enough here.

`Vec::iter` is used to iterate over a reference to vector.

Comment (kdy1) [reply crates/swc_ecma_parser/src/parser/expr.rs:795]: As it does not require ownership of the vector, you can remove `cloned()` above.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

File: crates/swc_ecma_parser/src/parser/expr.rs (hunk 3 above)
Comment (0xe) [reply crates/swc_ecma_parser/src/parser/expr.rs]: Thanks. I've removed cloned() and i'm accessing through reference.

**Pull request merged**